### PR TITLE
Pubdev 3454 glrm stalling

### PR DESCRIPTION
**Pull request by wendycwong**
Added code to increase GLRM run time by:

1. Remove repetitive operation from model iterations to outer loops.  The standardization operation is done at least three times per model iteration.  I move the standardization outside the loop.  We do it once before we go into model iteration and de-standardization once the model iteration is done.

2. Move the memory allocation to outer loops.  This is proposed by Pasha.

3. Access the data directly instead of trying to copy them into arrays.

I am running some profiling tests right now with customer data before I made the change and after I make the change to see what kind of improvements we are getting.

---
This change is [Reviewable](https://reviewable.io/reviews/h2oai/h2o-3/363)

### CODE REVIEW

File: h2o-algos/src/main/java/hex/glrm/GLRM.java [hunk 1]
```diff
@@ -1131,13 +1186,29 @@ private Chunk chk_xnew(Chunk[] chks, int c) {
 
           // Gradient wrt x_i is matrix product \grad L_{i,j}(x_i * Y_j, A_{i,j}) * Y_j'
           double[] weight = _lossFunc[j].mlgrad(xy, (int) a[j]);
```
Comment (st-pasha) [inline h2o-algos/src/main/java/hex/glrm/GLRM.java]: If you make `mlgrad` function to also accept the array to fill as an argument, then it would be possible to reuse `weight` as well.

Comment (st-pasha) [inline h2o-algos/src/main/java/hex/glrm/GLRM.java]: Moreover, you can add `len` (length of `xy`) as an additional parameter, then `xy` can be reused across iterations as well, provided that you initialize it to length `max(_yt.numLevels)`.

Comment (st-pasha) [inline h2o-algos/src/main/java/hex/glrm/GLRM.java]: These changes should improve performance, but make function `mlgrad` less readable. This can be offset however by providing detailed javadoc for that function, explaining why the arguments are the way they are, and what are the implicit assumptions.

File: h2o-algos/src/main/java/hex/glrm/GLRM.java [hunk 2]
```diff
@@ -1131,13 +1186,29 @@ private Chunk chk_xnew(Chunk[] chks, int c) {
...
-            for (int c = 0; c < weight.length; c++)
-              grad[k] += cweight * weight[c] * ysub[k][c];
+            for (int c = 0; c < weight.length; c++) {
+ //             grad[k] += cweight * weight[c] * ysub[k][c];
+              int cidx = _yt.getCatCidx(j, c);
+              double archtypevalues = 0;
+              if (_yt._transposed)
+                archtypevalues = _yt._archetypes[cidx][k];
+              else
+                archtypevalues = _yt._archetypes[k][cidx];
```
Comment (st-pasha) [inline h2o-algos/src/main/java/hex/glrm/GLRM.java]: just a nitpick, but seems like a ternary if would be more readable here:
`double archetypevalue = _yt._transposed? _yt._archetypes[cidx][k] : _yt._archetypes[k][cidx];`

File: h2o-algos/src/main/java/hex/glrm/GLRM.java [hunk 3]
```diff
@@ -1173,21 +1248,52 @@ private Chunk chk_xnew(Chunk[] chks, int c) {
         // Categorical columns
         for (int j = 0; j < _ncats; j++) {
           if (Double.isNaN(a[j])) continue;   // Skip missing observations in row
-          double[] xy = ArrayUtils.multVecArr(xnew, _yt.getCatBlock(j));
-          _loss +=  _lossFunc[j].mloss(xy, (int) a[j]);
+//          double[] xy = ArrayUtils.multVecArr(xnew, _yt.getCatBlock(j));
+//          _loss +=  _lossFunc[j].mloss(xy, (int) a[j]);
+
+          double[] txy = multVecArrFast(xnew, _yt, j);
+          _loss +=  _lossFunc[j].mloss(txy, (int) a[j]);
```
Comment (st-pasha) [inline h2o-algos/src/main/java/hex/glrm/GLRM.java]: Same comment as above applies here as well: right now `txy` is re-allocated on every iteration (and for every categorical). However the `xy` above can be reused here, provided that you add an extra parameter "effective length of `xy`" to the signature of `multVecArrFast`.

File: h2o-algos/src/main/java/hex/glrm/GLRM.java [hunk 4]
```diff
@@ -1316,6 +1427,82 @@ private Chunk chk_xnew(Chunk[] chks, int c) {
     }
   }
 
+  // Perform de-standardization over matrix A for non NA values only
+  private static class DeStandardizeA extends MRTask<DeStandardizeA> {
```
Comment (st-pasha) [inline h2o-algos/src/main/java/hex/glrm/GLRM.java]: Seems like this class is exactly the same as `StandardizeA`, only uses different `normSub`. If you create `deSub` array similarly how you had `deMul` above, then it should be possible to remove this class altogether

File: h2o-algos/src/main/java/hex/glrm/GLRM.java [hunk 5]
```diff
@@ -1316,6 +1427,82 @@ private Chunk chk_xnew(Chunk[] chks, int c) {
...
+      _ncolA = ncolA;
+      _ncolX = ncolX;
+      _ncats = ncats;
+
+      _normSub = normSub;
+      _normMul = normMul;
+    }
+
+    @SuppressWarnings("ConstantConditions")  // The method is too complex
+    @Override public void map(Chunk[] cs) {
```
Comment (st-pasha) [inline h2o-algos/src/main/java/hex/glrm/GLRM.java]: It appears that this method modifies the original data frame, which is problematic.
For instance, the user might want to do something with the frame while the model is being trained; or maybe the model terminates unexpectedly -- and then we never get to "denormalization" step -- in this case the use will see that running GLRM has "broken" their data.... Finally, there might be small round-off errors introduced during normalization/denormalization, which may matter for those pesky humans.

I would prefer that this method created modified copies of the original Vecs that need to be modified, and left the original Vecs intact.

**Comment by h2o-ops** [comment]
Tests passed!

File: h2o-algos/src/main/java/hex/glrm/GLRM.java [hunk 6]
```diff
@@ -625,18 +636,30 @@ public void computeImpl() {
...
 
         // Compute initial objective function
         _job.update(1, "Computing initial objective function");   // One unit of work
-        boolean regX = _parms._regularization_x != GlrmRegularizer.None && _parms._gamma_x != 0;  // Assume regularization on initial X is finite, else objective can be NaN if \gamma_x = 0
-        ObjCalc objtsk = new ObjCalc(_parms, yt, _ncolA, _ncolX, dinfo._cats, model._output._normSub,
-                                     model._output._normMul, model._output._lossFunc, weightId, regX);
+        // Assume regularization on initial X is finite, else objective can be NaN if \gamma_x = 0
+        boolean regX = _parms._regularization_x != GlrmRegularizer.None && _parms._gamma_x != 0;
+
+        // perform normalization on matrix A if needed.
```
Comment (tomasnykodym) [inline h2o-algos/src/main/java/hex/glrm/GLRM.java]: I am not sure StandardizeA is a good idea at all.  Do you have any performance measurements to back this up? We should measure this change carefully.

Standardization was made on the fly for some good reasons:
a) dataset is compressed, most datasets compressed 2 - 4x. Explicitly standardizing results in blowing up to doubles, so possible 2 - 4x increase in dataset size! And they will stay that way btw, there is no going back. Once double always double.

b) in theory at least, standardization cost is negligible compared to everything else, so saving on that is likely not gonna be detectable at all. I am saying in theory cause you never really know until you measure on real example. But I want good proof here that this helps on multiple datasets. And the speedup would have to be substantial to offset for the increase in used memory.

c) you loose sparsity. does not matter now because GLRM does not exploit data sparsity, but it could and then that would be huge slowdown (orders of magnitude easily).

If we do it, it would be safer (e.g. what happens in parallel grid search?) and better (you're expanding the data anyways and there is no going back, once doubles always doubles, also loosing precision ) to produce new standardized columns and remove them after. Also, instead of making custom StandardizeA task, I would make generic standardize data task in VecUtils or some place similar.

File: h2o-algos/src/main/java/hex/glrm/GLRM.java [hunk 7]
```diff
@@ -1131,13 +1186,29 @@ private Chunk chk_xnew(Chunk[] chks, int c) {
...
+ //         double[][] ysub = _yt.getCatBlock(j);
           for (int k = 0; k < _ncolX; k++) {
-            for (int c = 0; c < weight.length; c++)
-              grad[k] += cweight * weight[c] * ysub[k][c];
+            for (int c = 0; c < weight.length; c++) {
+ //             grad[k] += cweight * weight[c] * ysub[k][c];
+              int cidx = _yt.getCatCidx(j, c);
+              double archtypevalues = 0;
+              if (_yt._transposed)
+                archtypevalues = _yt._archetypes[cidx][k];
```
Comment (tomasnykodym) [inline h2o-algos/src/main/java/hex/glrm/GLRM.java]: It would probably pay off to transpose the matrix physically in the beginning to optimize on memory access:
archtypevalues = _yt._archetypes[cidx][k] is really slow

Comment (wendycwong) [inline h2o-algos/src/main/java/hex/glrm/GLRM.java]: This is a feature of GLRM.  The user is allowed to specify the archetypes either as it is or as its transposed.  I have no control over this.  Sorry cannot change this one.  

Note that I am actually not doing any transpose at this point.  I am just accessing the archetype with different indexing depending if the user wants it transposed or not.

Comment (tomasnykodym) [inline h2o-algos/src/main/java/hex/glrm/GLRM.java]: Why not? You can't make a copy and transpose it at the beginning of the map?
It's ok, it was just a suggestion, not really needed for this update.

File: h2o-algos/src/main/java/hex/glrm/GLRM.java (hunk 6 above)
Comment (tomasnykodym) [inline h2o-algos/src/main/java/hex/glrm/GLRM.java]: This looks like a decent speedup. Can you also measure without the physical standardization please? Also, can you record number of iterations?

Comment (tomasnykodym) [inline h2o-algos/src/main/java/hex/glrm/GLRM.java]: Also can you push your benchmarking script pls? :)

File: h2o-algos/src/main/java/hex/glrm/GLRM.java (hunk 7 above)
Comment (wendycwong) [inline h2o-algos/src/main/java/hex/glrm/GLRM.java]: Tomas:

We need to calculate the Gram matrix which is Y_transpose * Y or the other way round.  So, sometimes, we will have our archetypes in normal form and sometimes in the transposed form.  I do not fully understand this thingy yet and is hesitate to make the big change.  

However, I can make small tweaks without breaking the algorithm.  Once I understand fully what is going on, I will make better changes that you would prefer.  Please stay tune.

I would like to hear your comments even if it is not needed for this update.


Thanks,
Wendy

File: h2o-algos/src/main/java/hex/glrm/GLRM.java (hunk 6 above)
Comment (wendycwong) [inline h2o-algos/src/main/java/hex/glrm/GLRM.java]: Tomas:

The runtime number to the left is the one without any changes.  The numbers on the right is the one with changes to standardization and memory optimization.  I am running codes to tests run time only with standardization and then only with memory optimization to see which part actually helps most.  Please stay tuned for the results.  Both versions take about 13 iterations before ending.

I am going to check int the benchmarking script next.  You can actually run it on your own if you want.  They are called: runit_glrm_pubdev_3454_NOFEATURE_20runs.R, runit_glrm_pubdev_3454_NOFEATURE.R


Thanks,
Wendy

File: h2o-algos/src/main/java/hex/glrm/GLRM.java (hunk 4 above)
Comment (wendycwong) [inline h2o-algos/src/main/java/hex/glrm/GLRM.java]: True.  This is an easy fix.  It will be done.

Comment (wendycwong) [inline h2o-algos/src/main/java/hex/glrm/GLRM.java]: Done.

File: h2o-algos/src/main/java/hex/glrm/GLRM.java (hunk 2 above)
Comment (wendycwong) [inline h2o-algos/src/main/java/hex/glrm/GLRM.java]: Tomas also commented on this wanting to make it more efficient.  I moved the transposed check to outer loops.  Done.

File: h2o-algos/src/main/java/hex/glrm/GLRM.java (hunk 7 above)
Comment (wendycwong) [inline h2o-algos/src/main/java/hex/glrm/GLRM.java]: Moved the transpose check to outer loops.  Done.

File: h2o-algos/src/main/java/hex/glrm/GLRM.java (hunk 1 above)
Comment (wendycwong) [inline h2o-algos/src/main/java/hex/glrm/GLRM.java]: Pasha:

Like this suggestion.  However, this will involve changes across several classes/methods.  I want to postpone this to PUBDEV-3560: GLRM Efficiency improvement: avoid re-allocating arrays on every iteration.

Wendy

File: h2o-algos/src/main/java/hex/glrm/GLRM.java (hunk 6 above)
Comment (tomasnykodym) [inline h2o-algos/src/main/java/hex/glrm/GLRM.java]: Looks like the data is not in the S3. At least I just synced small data and I don't have it.

File: h2o-algos/src/main/java/hex/glrm/GLRM.java (hunk 3 above)
Comment (wendycwong) [inline h2o-algos/src/main/java/hex/glrm/GLRM.java]: Pasha:

Like this suggestion. However, this will involve changes across several classes/methods. I want to postpone this to PUBDEV-3560: GLRM Efficiency improvement: avoid re-allocating arrays on every iteration.

Wendy
